chalk-solve/src/clauses/builtin_traits/fn_family.rs

```diff
@@ -138,7 +140,7 @@ pub fn add_fn_trait_program_clauses<I: Interner>(
             }
             _ => Ok(()),
         },
-        TyData::Function(fn_val) => {
+        TyData::Function(fn_val) if fn_val.safety == Safety::Safe && !fn_val.variadic => {
```

Comment:
hmm it'd be nice if we had a more DRY way to encapsulate these two tests. We could leave for later, but basically I'm imagining if the "fn def" and "fn pointer" types embedded some common type like rustc's `FnSig` and could then invoke something like `fn_sig.implements_fn_traits()`

Reply:
Definitely agree! I'll open an issue as a reminder.